### PR TITLE
[pytx3] add expiremental_fetch command

### PR DESCRIPTION
**Pull request by InderH**
Create a new fetch2 command that downloads threat data via the new threat_updates endpoint.

### CODE REVIEW

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [inline pytx3/pytx3/cli/fetch2.py:44]: The argument parsing for "--threat-types" and "--additional-tags" is fixed in a later commit (d3b7431)

File: pytx3/pytx3/indicator.py
Comment (InderH) [inline pytx3/pytx3/indicator.py]: Forgot to remove this comment, it is removed in a later commit (fb823db)

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:10]: For my own edification, what's the difference between `from . import XXX` and `from .. import XXX`

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:42]: Can we change the comments here and across pytx to use less specific TE terminology and more general terminology for how we think about the data model. Some examples:
"Indicators" -> "Threats"
"descriptors" -> "opinions"
"privacy group" -> "collaboration"

Here i'd like to see something more like: "Only fetch updates for threats which the given app has added opinions about.

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:47]: "Only fetch updates for threats of the given type"

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:52]: "Only fetch updates for threats which some application has given these tags to"

**Review by jeberl — CHANGES_REQUESTED** [review]
Some changes

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:137]: Eventually, we're going to want to change this to a collaboration object right?

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [reply pytx3/pytx3/cli/fetch2.py:44]: fixed how? What does `nargs='+'` do?

File: pytx3/pytx3/indicator.py
Comment (jeberl) [inline pytx3/pytx3/indicator.py:35]: Good job giving an example

File: pytx3/pytx3/indicator.py
Comment (jeberl) [inline pytx3/pytx3/indicator.py:60]: Why do we join these and then later split them. Why not keep nested structure?

File: pytx3/pytx3/signal_type/signal_base.py
Comment (jeberl) [inline pytx3/pytx3/signal_type/signal_base.py:62]: What's this used for?

File: pytx3/pytx3/signal_type/signal_base.py
Comment (jeberl) [inline pytx3/pytx3/signal_type/signal_base.py:114]: Why does indicator state have int key but descriptor state has string key?

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:72]: why not start at `time.time()` instead of 0?

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [inline pytx3/pytx3/cli/fetch2.py:144]: Can we set the return type?

**Review by jeberl — COMMENTED** [review]
I'm having a hard time understanding what this file is used for. When is self.INDICATOR_TYPE set? What does self.TYPE_TAG refer to? Perhaps best to go through together

**Comment by jeberl** [comment]
In general this looks quite good. One major question though is how are you testing this? There's a number of test cases I'd like to see to show just that this pulls any data in a paginated way and stores/loads it correctly. In future PRs and in backend code 

I'd also like @bodnarbm to review as the python/git guru. I believe he was also the reviewer for davids original work.

File: pytx3/pytx3/indicator.py
Comment (InderH) [reply pytx3/pytx3/indicator.py:60]: Following the format used in descriptor.py.

Alternatively I can store the list directly and it would be saved into the file like "[123,456,789]".

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [reply pytx3/pytx3/cli/fetch2.py:10]: Basic version is:

`from . import XXX` means from the directory that houses this file, import the module defined in the file XXX. So here it would import the file XXX.py from `pytx3/pytx3/cli`

`from .. import xxx` means from the parent directory of the directory that houses this file, import the module defined in the file XXX. So here it would import the file XXX.py from `pytx3/pytx3`

`from ..dataset import XXX` means from the module dataset that inside the parent directory of the directory of this file, import the class or symbol XXX. So here it would import the class or symbol XXX that is defined in the file `pytx3/pytx3/dataset.py`

There's other details I'm probably missing, but that's a basic gist of its use here.

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [inline pytx3/pytx3/cli/fetch2.py:16]: I'm unsure what the final command name should be here, but for now, we could go with something like `ExperimentalFetchCommand` (and change the related cli invocation words to match) to make it clear this is not intended for production use yet.

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [reply pytx3/pytx3/cli/fetch2.py:42]: I like the direction of changes to terminology, but I think these should happen first on the ThreatExchange api side. To prevent cognitive mismatch when comparing the pytx3 documentation to the ThreatExchange documentation, I think it may be good to keep the existing terminology and then overall them together (or around the same time).

Just my thought.

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [reply pytx3/pytx3/cli/fetch2.py:72]: Alteratively, we could do something similar to `fetch.py` that uses

```python
# Cause first print to be after ~5 seconds
self.last_update_printed = now - self.PROGRESS_PRINT_INTERVAL_SEC + 5
```

File: pytx3/setup.py
Comment (bodnarbm) [inline pytx3/setup.py:38]: Per comment above, once this switches over to urllib, go ahead and remove the requests requirement.

File: pytx3/pytx3/indicator.py
Comment (bodnarbm) [reply pytx3/pytx3/indicator.py:60]: Since this is for translation to the CSV format and is the same approach used by SimpleDescriptorRollup.as_row in descriptor.py, I'm good with keeping the join method here. I'm not a fan of having to escape things in CSV, so if we can avoid having to escape the commas, I'm all for it.

**Review by bodnarbm — CHANGES_REQUESTED** [review]
Overall this direction looks good to me, especially given that this will continue to be in progress work along with other changes to the TE api.

My main request for changes center around removing the `requests` dependency and moving the API code into the `TE.py` module. I also suggest naming the command something along the lines of ExperimentalFetch to put users on notice that this is still in development. But any runtime way of warning users is fine with me.

Before merging this, can you verify that the cli tool parses pre-existing state files created by older versions of the tool. The changes you made to dataset storage should be fine and backwards compatible, but we should test that a dataset cache created by an older version of the tool will still process in this newer version using the exist fetch command. Just don't want to break users of the existing fetch.

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py:44]: nargs means take in arguments as a list. The "+" means expect at least 1 value if the argument flag is given (i.e. don't allow --threat-types to be an empty list)

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py:42]: @jeberl are you ok with punting on the terminology update as @bodnarbm mentioned?

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py:144]: yup

File: pytx3/pytx3/signal_type/signal_base.py
Comment (InderH) [reply pytx3/pytx3/signal_type/signal_base.py:62]: Place holder for the super class - if a child class doesn't implement process_indicator it defaults to false.

File: pytx3/pytx3/signal_type/signal_base.py
Comment (InderH) [reply pytx3/pytx3/signal_type/signal_base.py:129]: indicator_state is keyed by indicator id, state (for descriptors) is keyed by the indicator (only 1 descriptor is stored per indicator)

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py:72]: Used 0 as an initial value so we would print right after the first fetch so the user gets some immediate feedback (so they know the command is doing stuff) vs having to wait 30s before the feedback.

Could do something similar to fetch.py if you prefer but I don't think it makes much of a difference

**Review by bodnarbm — COMMENTED** [review]
LGTM.

I'll leave the remainder to @jeberl to review if his items are fully addressed, but the changes you made @InderH resolved my prior comments. 

**My only additional ask is to please run the `black` formatter on the files in `pytx3` to make sure the code format is consistent with the existing files.**

```shell
$ black pytx3/**/*.py
```

assuming you have black installed with `pip install black`

Re testing, I tested locally that an existing fetch checkpoint and dataset from the version in master was able to be processed with the `fetch` command. My test plan was as follows.

```
❯ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.
❯ pip uninstall -y pytx3 && pip install -e pytx3
Found existing installation: pytx3 0.0.8
Uninstalling pytx3-0.0.8:
  Successfully uninstalled pytx3-0.0.8
Obtaining file:///home/bodnarbm/repos/facebook/ThreatExchange/pytx3
Requirement already satisfied: python-Levenshtein in /home/bodnarbm/.venv/threatexchange/lib/python3.8/site-packages (from pytx3==0.0.8) (0.12.0)
Requirement already satisfied: setuptools in /home/bodnarbm/.venv/threatexchange/lib/python3.8/site-packages (from python-Levenshtein->pytx3==0.0.8) (44.0.0)
Installing collected packages: pytx3
  Running setup.py develop for pytx3
Successfully installed pytx3
❯ pytx3 fetch --clear
❯ pytx3 fetch --until-timestamp $TIMESTAMP_1
photo_md5: 1
video_md5: 4
❯ pytx3 fetch --until-timestamp $TIMESTAMP_2
Doing an incremental update from ****************
This will miss some updates and deletes of previously fetched data, use --full to force a refetch.
photo_md5: 4
pdq: 5
video_md5: 877
❯ pytx3 fetch --clear
❯ pytx3 fetch --until-timestamp $TIMESTAMP_1
photo_md5: 1
video_md5: 4
❯ git checkout InderH/fetch2
Switched to branch 'InderH/fetch2'
❯ pip uninstall -y pytx3 && pip install -e pytx3
Found existing installation: pytx3 0.0.8
Uninstalling pytx3-0.0.8:
  Successfully uninstalled pytx3-0.0.8
Obtaining file:///home/bodnarbm/repos/facebook/ThreatExchange/pytx3
Requirement already satisfied: python-Levenshtein in /home/bodnarbm/.venv/threatexchange/lib/python3.8/site-packages (from pytx3==0.0.8) (0.12.0)
Requirement already satisfied: setuptools in /home/bodnarbm/.venv/threatexchange/lib/python3.8/site-packages (from python-Levenshtein->pytx3==0.0.8) (44.0.0)
Installing collected packages: pytx3
  Running setup.py develop for pytx3
Successfully installed pytx3
❯ pytx3 fetch --until-timestamp $TIMESTAMP_2
Doing an incremental update from ************
This will miss some updates and deletes of previously fetched data, use --full to force a refetch.
photo_md5: 4
video_md5: 877
pdq: 5
```

I don't see any regression, so this looks safe to merge from a production code path viewpoint.

Note: I did not test the experimental fetch itself, and I leave that to you and @jeberl on what testing is needed for the experimental path given planned changes on the API end of the work.

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [reply pytx3/pytx3/cli/fetch2.py]: I should note that my comment was also specific to across pytx3. For this command, given it's intended use, I'm less concerned about the changes. And I'm fine either way, but would prefer to defer the terminology change across pytx3 until we are ready to make that change in the API layer (to prevent potential churn in exact wording).

File: pytx3/pytx3/cli/fetch2.py
Comment (bodnarbm) [reply pytx3/pytx3/cli/fetch2.py]: Either way is fine with me. We can iterate on the UX for the command too once it is out of experimental.

**Review by bodnarbm — APPROVED** [review]

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [reply pytx3/pytx3/cli/fetch2.py]: genius. thank you

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [reply pytx3/pytx3/cli/fetch2.py]: @InderH, I'll leave to you

File: pytx3/pytx3/cli/fetch2.py
Comment (jeberl) [reply pytx3/pytx3/cli/fetch2.py]: Ok. I'm taking a documentation refresh on as a project this half so can come back and make these edits

File: pytx3/pytx3/indicator.py
Comment (jeberl) [reply pytx3/pytx3/indicator.py:60]: sgtm

File: pytx3/pytx3/cli/experimental_fetch.py
Comment (jeberl) [inline pytx3/pytx3/cli/experimental_fetch.py:14]: Delete extra line

**Review by jeberl — COMMENTED** [review]
Left a bunch few comments through out and Resolved the comments that I think you fix. Once those are set should be good.

Still getting the hang of how to review via git

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py]: Function no longer exists; if you look at the newest changes I moved it the fetching into TE.py

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py]: Yup, threat_updates defaults it to request.time

File: pytx3/pytx3/cli/fetch2.py
Comment (InderH) [reply pytx3/pytx3/cli/fetch2.py]: Added into task 2.10 of my milestone 2 plan, I can send you a link if you'd like.

File: pytx3/pytx3/TE.py
Comment (InderH) [reply pytx3/pytx3/TE.py:180]: The actual `experimentalfetch` command is ~37% of the loc of `fetch`, granted this may change a little bit as I update experimentalfetch once threat_updates is updated.

The code is also much simpler as threat_updates is a much more direct way to download data compared to what `fetch` does.

Comment (jeberl) [reply pytx3/pytx3/TE.py:180]: Incredible!!!

**Review by jeberl — APPROVED** [review]

**Comment by jeberl** [comment]
Great stuff! Your first GitHub PR at fb!

Merge it...

**Pull request merged**
